snapshot/snapshotter.go

```diff
@@ -253,82 +258,91 @@ func makeTest(t *testing.T, name string, snapshotterFn func(root string) (Snapsh
 	}
 }
 
+func tFatal(t *testing.T, err error) {
+	t.Fatal(errors.WithStack(err))
+}
+
 // checkSnapshotterBasic tests the basic workflow of a snapshot snapshotter.
 func checkSnapshotterBasic(t *testing.T, snapshotter Snapshotter, work string) {
+	initialApplier := fstest.Apply(
+		fstest.CreateFile("/foo", []byte("foo\n"), 0777),
+		fstest.CreateDir("/a", 0755),
+		fstest.CreateDir("/a/b", 0755),
+		fstest.CreateDir("/a/b/c", 0755),
+	)
+
+	diffApplier := fstest.Apply(
+		fstest.CreateFile("/bar", []byte("bar\n"), 0777),
+		// also, change content of foo to bar
+		fstest.CreateFile("/foo", []byte("bar\n"), 0777),
+		fstest.RemoveAll("/a/b"),
+	)
+
 	preparing := filepath.Join(work, "preparing")
 	if err := os.MkdirAll(preparing, 0777); err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 
 	mounts, err := snapshotter.Prepare(preparing, "")
 	if err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 
 	if len(mounts) < 1 {
-		t.Fatal("expected mounts to have entries")
+		tFatal(t, errors.New("expected mounts to have entries"))
 	}
 
 	if err := containerd.MountAll(mounts, preparing); err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 	defer testutil.Unmount(t, preparing)
 
-	if err := ioutil.WriteFile(filepath.Join(preparing, "foo"), []byte("foo\n"), 0777); err != nil {
-		t.Fatal(err)
-	}
-
-	if err := os.MkdirAll(filepath.Join(preparing, "a", "b", "c"), 0755); err != nil {
-		t.Fatal(err)
+	if err := initialApplier.Apply(preparing); err != nil {
+		tFatal(t, err)
 	}
 
 	committed := filepath.Join(work, "committed")
 	if err := snapshotter.Commit(committed, preparing); err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 
 	si, err := snapshotter.Stat(committed)
 	if err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 
 	assert.Equal(t, si.Parent, "")
 
 	next := filepath.Join(work, "nextlayer")
 	if err := os.MkdirAll(next, 0777); err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 
 	mounts, err = snapshotter.Prepare(next, committed)
 	if err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 	if err := containerd.MountAll(mounts, next); err != nil {
-		t.Fatal(err)
+		tFatal(t, err)
 	}
 	defer testutil.Unmount(t, next)
 
-	if err := ioutil.WriteFile(filepath.Join(next, "bar"), []byte("bar\n"), 0777); err != nil {
-		t.Fatal(err)
-	}
-
-	// also, change content of foo to bar
-	if err := ioutil.WriteFile(filepath.Join(next, "foo"), []byte("bar\n"), 0777); err != nil {
-		t.Fatal(err)
+	if err := initialApplier.Verify(next); err != nil {
+		tFatal(t, err)
 	}
 
-	if err := os.RemoveAll(filepath.Join(next, "a", "b")); err != nil {
-		t.Log(err)
+	if err := diffApplier.Apply(work); err != nil {
```

Comment:
Shouldn't this be applied onto `next`? Also would expect a directory check post commit.